### DISCUSSION TITLE
Проверка на уникальность при добавлении + regexp

### DISCUSSION THREAD

**Discussion by Qb1cc** [totumonline/totum-ru-issues-and-discussions#119] [Как это сделать?] [answered]
Нужно сделать проверку уникальности перед сохранением в таблицу товара из цикла, что бы не плодились дубли в таблице товаров, Как правильно это сделать? + еще нужно удалять все символы и пробелы с артикула. 

По идее нужно сделать сначала форматирование с удалением ненужного => проверку на дубли по таблице => сохранение.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Добавляете код действия при добавлении, который это проверяет и отфутболивает если добавляется дубль:

```
ad1= : if(condition: $sel != ""; then: $err)

	err: errorException(text: "This item is already on this order!")

sel: select(table: $#ntn; field: 'id'; where: 'item' = $#nfv; where: 'id' != #id)
```

**Reply by Qb1cc** [reply]
Это я видел в примерах. Мне не понятно как удалять символы при добавлении из введённых значений. 
В поле артикул обычно менеджеры копипастом добавляют, но копированное значение может быть разных форматов с пробелами, дефисами и т.д. Нужно не просто проверять значение и выводить ошибку , а сразу изменять его так чтобы оставались только буквы и числа ( подгонять в общем под общий формат артикулы ) => потом проверка уже на дубли и сохранение.  Я пробывал использовать StrReplace , в стр что то не то передаю видимо

**Reply by ttm-alexeypolunin** [reply]
Да, не увидел. Ща будет гениально (это настоящий программист придумал):

```
// Код действия при добавлении

ad1= : if(condition: $regm = true; then: $set) 

	set: set(table: $#ntn; field: $#nf = $replace; where: 'id' = #id)

~regm: strRegAllMatches(str: $#nfv; template: "[0-9]+"; matches: "matches"; flags: "u")

replace: listJoin(list: $#matches[0]; str: "")

```

Но вам еще один `condition` в `if` надо воткнуть с проверкой на дубль.

**Reply by Qb1cc** [reply]
Да всё отлично заработало, больше спасибо. Единственное изменил выражение `[0-9]+`  на `[^!@#$%&*( )_-]`  что бы , удаляло пробелы и указанные символы.

**Reply by Qb1cc** [reply]
Всем читающим, гит. Что бы работал ad1 , ad2  корректно( или нет, главное работает ) в этом случае. Нужно поменять на a1, a2  ( объяснение есть в учебнике ) 
Но тут тоже есть баги с точкой , ну или я не правильное выражение сделал. Если руки дойдут, сделаю видео, это нужно показывать именно. Баги проявляются только с точкой и коротких артикулах 2-3 знака. Встретить в реальной жизни такое не реально, так что можно использовать, а даже если и встретите то база от этого не сильно пострадает.
```
//Удаляем символы указанные в template ( изучайте регулярные выражения ) 
a1= : if(condition: $regm = true; then: $set) 

	set: set(table: $#ntn; field: $#nf = $replace; where: 'id' = #id)

~regm: strRegAllMatches(str: $#nfv; template: "[^.!#*( ),_-]"; matches: "matches"; flags: "u")

replace: listJoin(list: $#matches[0]; str: "")

//Проверяем на дубли 
a2= : if(condition: $sel != ""; then: $err)

	err: errorException(text: "Такой артикул уже есть в базе!")

sel: select(table: "$#ntn"  ; field: 'id'; where: 'ваш name поля' = $#nfv; where: 'id' != #id)
```

**Reply by ttm-alexeypolunin** [reply]
Пример как ОТРЕЗАТЬ все символы, кроме разрешенных:

```
=: if(condition: $matches = true; then: $replace; else: #some_data)

matches: strRegAllMatches(str: #some_data; template: "([^a-zA-Z0-9а-яА-Я .,\-+{}()]+)"; matches: "matches"; flags: "u")

replace: strReplace(str: #some_data; from: $#matches[0]; to: "")
```

Отрезаем все, что не `a-zA-Z0-9а-яА-Я .,\-+{}()`

**Comment by mihail-cornev** [comment]
Подскажите, а почему выдаёт ошибку или как убрать текст ошибки?

![Screenshot_1](https://user-images.githubusercontent.com/113771444/202640728-5e825c17-47ac-4f39-8107-b0a096931403.jpg)

и второй вопрос
как можно проверять на совпадение, не учитывая регистр.

чтобы такое считалось как одинаковый
KP30-NA-600
kp30-na-600

**Reply by erpic-ru** [reply]
Технический текст ошибки отображается только у администратора. У пользователя его нет

Нужно техническое поле с функцией [strtransform "lower"](https://ru.docs.totum.online/functions#fn-strTransform) и сравнивать по нему

**Reply by alexrus79** [reply]
Использовать флаг "i" в функции strRegAllMatches
